src/fsharp/CompileOps.fs

```diff
@@ -3949,14 +3951,25 @@ and [<Sealed>] TcImports(tcConfigP: TcConfigProvider, initialResolutions: TcAsse
 #endif
 
     let disposal = new TcImportsSafeDisposal(disposeActions, disposeTypeProviderActions, compilationThread)
-    
+
     let CheckDisposed() =
         if disposed then assert false
 
     let dispose () =
         CheckDisposed()
         (disposal :> IDisposable).Dispose()
 
+    let fixupOrphanCcus () =
+        let keys = ccuThunks.Keys
```

Comment:
Iterating ConcurrentDictionary Keys while modifying the concurrent dictionary seems a bit dodgy.  I read something of what happens here: https://stackoverflow.com/questions/10479867/is-concurrentdictionary-keys-or-values-property-threadsafe

What if the process of fixing things up creates more entries that need to be fixed up?  Should this iterate until no more fixups have occurred?

Reply:
@dsyme, I will take a look at it, but basically, I just need a snapshot.  I assumed, that the keys and values collection on would do that.  Given the durned thing is supposed to be a threadsafe dictionary.  I remove an entry from the collection to fix it up, to stop any other threads from doing so, then I put it back, if fixup failed to achieve anything usefull.

Reply:
@dsyme, yes I was right it gets a snapshot.  However, I will add a comment describing what this does, because it is a touch fruity.

Reply:
Added description in comment.